### DISCUSSION TITLE
Sending Marantz in Pronto format

### DISCUSSION THREAD

**Discussion by elockman** [Arduino-IRremote/Arduino-IRremote#1038] [Help] [answered]
I am running into an issue where IR commands stop working after the first transmission.

I have a simple program to turn a Marantz AVR on and off as a test, every 6 seconds.

```
void loop(){
  hex_marantz_on();
  delay(6000);
  hex_marantz_off();
  delay(6000);  
}
```
Only the first IR command executes.  If I swap commands and Flash, the AVR will switch power states once.

These are the commands I'm using:
```
void hex_marantz_on()
{
  irsend.sendPronto("0000 0071 0000 0024 0020 0020 0020 0020 0020 0020 0040 0020 0020 0020 0020 0020 0020 00A1 0020 0020 0020 0040 0020 0020 0040 0020 0020 0020 0020 0020 0020 0020 0020 0020 0020 0020 0020 0040 0020 0AAB 0020 0020 0020 0020 0020 0020 0040 0020 0020 0020 0020 0020 0020 00A1 0020 0020 0020 0040 0020 0020 0040 0020 0020 0020 0020 0020 0020 0020 0020 0020 0020 0020 0020 0040 0020 04B0 ", IR_REPEATS);
}

void hex_marantz_off()
{
  irsend.sendPronto("0000 0071 0000 0022 0020 0020 0020 0020 0020 0020 0040 0020 0020 0020 0020 0020 0020 00A1 0020 0020 0020 0040 0020 0020 0040 0020 0020 0020 0020 0020 0020 0020 0020 0020 0020 0040 0040 0AAB 0020 0020 0020 0020 0020 0020 0040 0020 0020 0020 0020 0020 0020 00A1 0020 0020 0020 0040 0020 0020 0040 0020 0020 0020 0020 0020 0020 0020 0020 0020 0020 0040 0040 04B0 ", IR_REPEATS);
}
```
IR_REPEATS is 3, but I've tried reducing all the way down to 0.

Next, I tried just to run a simple power toggle program
```
void hex_marantz_on_off()
{
//  toggle power
  irsend.sendPronto("0000 0073 0000 0018 0020 0020 0020 0020 0020 0020 0040 0020 0020 0020 0020 0020 0020 0020 0020 0020 0020 0040 0020 0020 0040 0020 0020 0C5B 0020 0020 0020 0020 0020 0020 0040 0020 0020 0020 0020 0020 0020 0020 0020 0020 0020 0040 0020 0020 0040 0020 0020 0498 ", IR_REPEATS);
}
void loop()
{
  hex_marantz_on_off();
  delay(6000);
}
```
Only the first IR command works.  I must cycle IR device power to toggle the AVR power.
Anyone have an idea why the device needs to be rebooted? Can I reset the the IR buffers once sending the commands?

**Comment by elockman** [comment]
The IR codes above came directly from Marantz's documentation. They are labeled "Hex Code (Basic Commands Only)" 

They work (kind of), but do they have to be converted to Pronto somehow?

**Comment by ArminJo** [comment]
@bengtmartensson Any Idea?

**Comment by bengtmartensson** [comment]
The given signal decode (in IrScrutinizer) as RC5x: {D=16,S=12,F=1,T=1} and RC5x: {D=16,S=12,F=2,T=1}, in both cases two copies. The RC5x is a toggling protocol, meaning that the toggle ("T" above") is supposed to toggle between 0 and 1 on subsequent invocations. But you are sending using T=1 all of the time, so the receiver ignores all but the first one, as it should. You need to send signals with T=0 alternating.

Corrsponding signals with T=0 can be generated with IrScrutinizer:

{D=16,S=12,F=1,T=0}: 0000 0073 0000 0011 0020 0020 0040 0040 0040 0020 0020 0020 0020 0020 0020 00A0 0020 0020 0020 0040 0020 0020 0040 0020 0020 0020 0020 0020 0020 0020 0020 0020 0020 0020 0020 0040 0020 0AA8

{D=16,S=12,F=2,T=0}: 0000 0073 0000 0010 0020 0020 0040 0040 0040 0020 0020 0020 0020 0020 0020 00A0 0020 0020 0020 0040 0020 0020 0040 0020 0020 0020 0020 0020 0020 0020 0020 0020 0020 0040 0040 0AC8

So if you change your code to something like:

```
static unsigned int T = 0U;
...
void hex_marantz_on()
{
    if (T)
        irsend.sendPronto("0000 0073 0000 0012 0020 0020 0020 0020 0020 0020 0040 0020 0020 0020 0020 0020 0020 00A0 0020 0020 0020 0040 0020 0020 0040 0020 0020 0020 0020 0020 0020 0020 0020 0020 0020 0020 0020 0040 0020 0AA8", IR_REPEATS);
    else
        irsend.sendPronto("0000 0073 0000 0011 0020 0020 0040 0040 0040 0020 0020 0020 0020 0020 0020 00A0 0020 0020 0020 0040 0020 0020 0040 0020 0020 0020 0020 0020 0020 0020 0020 0020 0020 0020 0020 0040 0020 0AA8", IR_REPEATS):

    T = 1U - T;
}
```

it should work. 

BTW, Infrared4Arduino handles the sending with alternating toggles automatically (for the RC5 protocol).

**Reply by ArminJo** [reply]
@bengtmartensson Thanks for the explanation.
BTW, the Irremote function ~~sendRC5ext(uint8_t addr, uint8_t cmd, bool toggle)~~ also supports automatic toggling 😀 
Update: **sendRC5(uint8_t aAddress, uint8_t aCommand, int_fast8_t aNumberOfRepeats, bool aEnableAutomaticToggle)** supports automatic toggle.

**Comment by elockman** [comment]
@bengtmartensson thank you! I am new to IR, but I downloaded IrScrutinizer and it's a huge help.

@ArminJo if I get the following result from IrScrutinizer, how do I assign the addr and cmd in sendRC5ext()?
```
RC5x: {D=16,F=2,S=12,T=1}, beg=0, end=67, reps=2
{36.7k,872,msb}<1,-1|-1,1>(A:8,-4,B:12,-73m,C:3,2,D:3,-5,E:12,-31.841m)*{A=0xf0,B=0x302,C=0x0,D=0x7,E=0x302}
```

**Reply by elockman** [reply]
@ArminJo I got it working. addr=16 (D), cmd=12 (S)

It is working with:
`irsend.sendRC5ext(16, 12, false);
`

**Reply by elockman** [reply]
@ArminJo I'm looking closer at the code. It appears sendRC5ext() is deprecated, due to sendRC5() handling both protocols.

However, RC5x should also accept an extension argument (F=?). Is that supported in this library?

**Reply by ArminJo** [reply]
What is an extension argument???

**Reply by ArminJo** [reply]
OK, it is Marantz specific and not covered by RC5 definitions :-(.
You can try to receive the commands with IRremote, to look, what these special extensions are encoded to.

**Comment by elockman** [comment]
@ArminJo Below is a screenshot of the Marantz document I am referencing.

For clarity, I want to send a **POWER OFF** command rather than an **POWER ON/OFF** command.  This keeps the AVR from turning on by inadvertently toggling from the off state.

The following yields a toggle (which makes sense to me, since I cannot define the 02 extension) 
`irsend.sendRC5ext(16, 12, false);`
The document indicates I should send addr=16, cmd=12, and ext=2 for **POWER OFF.**

@bengtmartensson I assumed the extension argument is the difference between RC5 and RC5ext protocols? 

![Screenshot_20221101_134107](https://user-images.githubusercontent.com/8911455/199323174-e4ad3af1-3a55-41d8-9867-ea155621559c.png)

Full document can be downloaded at Marantz website (Note there are 2 spreadsheet tabs)
https://www.marantz.com/-/media/files/documentmaster/marantzna/us/marantz_fy21_sr8015_sr7015_ir_code_v01.xls

**Comment by bengtmartensson** [comment, marked as answer]
Most protocols/manufacturer (NEC being the canonical example) call their parameters device (D), subdevice (S) and function(F). Philips (creator and main user of the RC5 protocol is a tad different: By RC5, the are calling the "device" parameter "system" instead, and said "command" for "function". When the RC5 protocol was going too small, they did not create a "subdevice", but extended the command/function instead. See above: command 12 is power toggle by very many RC5 devices; it is "extended" with a "extension" to form discrete power on and off commands. 

IrScrutinizer (which builds on IrpTransmogrifer) borrows the definitions of RC5, RC5x, RC6* etc from the JP1-community, in particular from DecodeIR. In RC5x, it calls system for D, command for "S", and extension for F.  This is not really cool ... :-\

**Reply by elockman** [reply]
Thanks again, @bengtmartensson!

I finally have it working now. I've tested by sending the command multiple times to make sure it goes off, but does not come on with toggle. (I am toggling between T=0 and T=1)

I used **IrScrutinizer** to calculate the outputs.
The Pronto and RC5 protocols did not work, but I was able to get it to work by sending raw data.
```
//RC5x: {D=16,F=2,S=12,T=1}, beg=0, end=33, reps=1
//0000 0073 0000 0011 0020 0020 0020 0020 0020 0020 0040 0020 0020 0020 0020 0020 0020 00A0 0020 0020 0020 0040 0020 0020 0040 0020 0020 0020 0020 0020 0020 0020 0020 0020 0020 0040 0040 0AC8
  const uint16_t offRawData[] = {889,889,889,889,889,889,1778,889,889,889,889,889,889,4445,889,889,889,1778,889,889,1778,889,889,889,889,889,889,889,889,889,889,1778,1778};
  irsend.sendRaw(offRawData, sizeof(offRawData) / sizeof(offRawData[0]), RC5_RC6_KHZ);
  delay(50); // delay must be greater than 5 ms (RECORD_GAP_MICROS), otherwise the receiver sees it as one long signal

```
